### PR TITLE
Improved code

### PR DESCRIPTION
**Pull request by sayem314**
Followup: https://github.com/Angristan/OpenVPN-install/pull/208

### CODE REVIEW

**Comment by cezar97** [comment]
I apologise, but if it's a follow-up it shouldn't have the same changes. Just leave one with only "improved codes" part and another with only support for custom DNS. Requested in #208, too.
Make a new branch from master and apply changes, or if you want to use the Github UI click the edit icon on Angristan repo and it automatically fork and make a new branch for you.
Thank you for the PRs anyway ;-).

**Comment by sayem314** [comment]
This is not same changes to #208. It is only improved codes. I was thinking to PR dns one later. But no one gives a fuck actually it seems.

**Comment by cezar97** [comment]
I'd like to see custom DNS merged.
> But no one gives a fuck actually it seems.

Yeah, you're right... Sorry @Angristan. But yes, we're in a PR freeze... Sorry for this @sayem314.

File: openvpn-install.sh [hunk 1]
```diff
@@ -22,7 +25,7 @@ fi
 if [[ -e /etc/debian_version ]]; then
 	OS="debian"
 	# Getting the version number, to verify that a recent version of OpenVPN is available
-	VERSION_ID=$(cat /etc/os-release | grep "VERSION_ID")
+	VERSION_ID=$(grep "VERSION_ID" < /etc/os-release)
```
Comment (cezar97) [inline openvpn-install.sh]: `<` is not needed with `grep`.

Comment (sayem314) [reply openvpn-install.sh]: I see. This can be removed later if PR get merged.

File: openvpn-install.sh [hunk 2]
```diff
@@ -59,28 +62,33 @@ fi
...
-	echo "</key>" >> $homeDir/$1.ovpn
-	echo "key-direction 1" >> $homeDir/$1.ovpn
-	echo "<tls-auth>" >> $homeDir/$1.ovpn
-	cat /etc/openvpn/tls-auth.key >> $homeDir/$1.ovpn
-	echo "</tls-auth>" >> $homeDir/$1.ovpn
+	cp /etc/openvpn/client-template.txt "$homeDir/$1.ovpn"
+	{
+		echo "<ca>"
+		cat "/etc/openvpn/easy-rsa/pki/ca.crt"
+		echo "</ca>"
```
Comment (cezar97) [inline openvpn-install.sh]: Quotes not really needed with `cat` here.

File: openvpn-install.sh [hunk 3]
```diff
@@ -261,8 +280,8 @@ else
 	echo "   5) CAMELLIA-192-CBC"
 	echo "   6) CAMELLIA-256-CBC"
 	echo "   7) SEED-CBC"
-	while [[ $CIPHER != "1" && $CIPHER != "2" && $CIPHER != "3" && $CIPHER != "4" && $CIPHER != "5" && $CIPHER != "6" && $CIPHER != "7" ]]; do
-		read -p "Cipher [1-7]: " -e -i 1 CIPHER
+	until [[ "$DNS" =~ ^[0-9]+$ ]] && [ "$DNS" -ge 1 -a "$DNS" -le 7 ]; do
```
Comment (cezar97) [inline openvpn-install.sh]: There's a typo. It should be `$CIPHER` instead of `$DNS`. Also `0-7` instead of `0-9`.

File: openvpn-install.sh (hunk 2 above)
Comment (sayem314) [reply openvpn-install.sh]: Yeah. But I did to for making it look nicer since other cat below have quotes because of $1

File: openvpn-install.sh [hunk 4]
```diff
@@ -292,8 +311,8 @@ else
 	echo "   1) 2048 bits (fastest)"
 	echo "   2) 3072 bits (recommended, best compromise)"
 	echo "   3) 4096 bits (most secure)"
-	while [[ $DH_KEY_SIZE != "1" && $DH_KEY_SIZE != "2" && $DH_KEY_SIZE != "3" ]]; do
-		read -p "DH key size [1-3]: " -e -i 2 DH_KEY_SIZE
+	until [[ "$DNS" =~ ^[0-9]+$ ]] && [ "$DNS" -ge 1 -a "$DNS" -le 3 ]; do
```
Comment (cezar97) [inline openvpn-install.sh]: Same as my previous comment ( typo ).

File: openvpn-install.sh [hunk 5]
```diff
@@ -311,8 +330,8 @@ else
 	echo "   1) 2048 bits (fastest)"
 	echo "   2) 3072 bits (recommended, best compromise)"
 	echo "   3) 4096 bits (most secure)"
-	while [[ $RSA_KEY_SIZE != "1" && $RSA_KEY_SIZE != "2" && $RSA_KEY_SIZE != "3" ]]; do
-		read -p "RSA key size [1-3]: " -e -i 2 RSA_KEY_SIZE
+	until [[ "$DNS" =~ ^[0-9]+$ ]] && [ "$DNS" -ge 1 -a "$DNS" -le 3 ]; do
```
Comment (cezar97) [inline openvpn-install.sh]: Typo here too.

Comment (sayem314) [reply openvpn-install.sh]: 😭😭

**Comment by sayem314** [comment]
I will close this now. Will PR a fix tomorrow.

File: openvpn-install.sh (hunk 4 above)
Comment (sayem314) [reply openvpn-install.sh]: Fucking Ctrl+C,V

File: openvpn-install.sh (hunk 3 above)
Comment (cezar97) [reply openvpn-install.sh]: Yeah, it happens all time... :smile:

**Comment by cezar97** [comment]
You can leave it open, too.
I'll just review as "requested changes" because of typos, not for the other suff.

**Comment by sayem314** [comment]
Thanks! Reopened.

**Review by ghost — CHANGES_REQUESTED** [review]
Typo

File: openvpn-install.sh [hunk 6]
```diff
@@ -280,7 +280,7 @@ else
 	echo "   5) CAMELLIA-192-CBC"
 	echo "   6) CAMELLIA-256-CBC"
 	echo "   7) SEED-CBC"
-	until [[ "$DNS" =~ ^[0-9]+$ ]] && [ "$DNS" -ge 1 -a "$DNS" -le 7 ]; do
+	until [[ "$CIPHER" =~ ^[0-9]+$ ]] && [ "$CIPHER" -ge 1 -a "$CIPHER" -le 7 ]; do
```
Comment (cezar97) [inline openvpn-install.sh]: Regex should be `^[0-7]+$`?

File: openvpn-install.sh [hunk 7]
```diff
@@ -311,7 +311,7 @@ else
 	echo "   1) 2048 bits (fastest)"
 	echo "   2) 3072 bits (recommended, best compromise)"
 	echo "   3) 4096 bits (most secure)"
-	until [[ "$DNS" =~ ^[0-9]+$ ]] && [ "$DNS" -ge 1 -a "$DNS" -le 3 ]; do
+	until [[ "$DH_KEY_SIZE" =~ ^[0-9]+$ ]] && [ "$DH_KEY_SIZE" -ge 1 -a "$DH_KEY_SIZE" -le 3 ]; do
```
Comment (cezar97) [inline openvpn-install.sh]: Regex should be ^[0-3]+$?

File: openvpn-install.sh [hunk 8]
```diff
@@ -330,7 +330,7 @@ else
 	echo "   1) 2048 bits (fastest)"
 	echo "   2) 3072 bits (recommended, best compromise)"
 	echo "   3) 4096 bits (most secure)"
-	until [[ "$DNS" =~ ^[0-9]+$ ]] && [ "$DNS" -ge 1 -a "$DNS" -le 3 ]; do
+	until [[ "$RSA_KEY_SIZE" =~ ^[0-9]+$ ]] && [ "$RSA_KEY_SIZE" -ge 1 -a "$RSA_KEY_SIZE" -le 3 ]; do
```
Comment (cezar97) [inline openvpn-install.sh]: Regex should be ^[0-3]+$?

**Comment by sayem314** [comment]
@cezar97 this is not really a problem. This 0-9 regex means “take numbers only”. After this regex and “&&” actual value is verified.

**Comment by cezar97** [comment]
Right, thanks.

**Review by ghost — APPROVED** [review]

File: openvpn-install.sh [hunk 9]
```diff
@@ -242,8 +261,8 @@ else
 	echo "   7) Google (Anycast: worldwide)"
 	echo "   8) Yandex Basic (Russia)"
 	echo "   9) AdGuard DNS (Russia)"
-	while [[ $DNS != "1" && $DNS != "2" && $DNS != "3" && $DNS != "4" && $DNS != "5" && $DNS != "6" && $DNS != "7" && $DNS != "8" && $DNS != "9" ]]; do
-		read -p "DNS [1-9]: " -e -i 1 DNS
+	until [[ "$DNS" =~ ^[0-9]+$ ]] && [ "$DNS" -ge 1 -a "$DNS" -le 9 ]; do
```
Comment (angristan) [inline openvpn-install.sh:264]: Why the `-a` here? 🤔

File: openvpn-install.sh [hunk 10]
```diff
@@ -311,8 +330,8 @@ else
 	echo "   1) 2048 bits (fastest)"
 	echo "   2) 3072 bits (recommended, best compromise)"
 	echo "   3) 4096 bits (most secure)"
-	while [[ $RSA_KEY_SIZE != "1" && $RSA_KEY_SIZE != "2" && $RSA_KEY_SIZE != "3" ]]; do
-		read -p "RSA key size [1-3]: " -e -i 2 RSA_KEY_SIZE
+	until [[ "$RSA_KEY_SIZE" =~ ^[0-9]+$ ]] && [ "$RSA_KEY_SIZE" -ge 1 -a "$RSA_KEY_SIZE" -le 3 ]; do
```
Comment (angristan) [inline openvpn-install.sh:333]: Why the `+$` here?

File: openvpn-install.sh [hunk 11]
```diff
@@ -494,11 +515,7 @@ WantedBy=multi-user.target" > /etc/systemd/system/iptables.service
 
 	# Generate server.conf
 	echo "port $PORT" > /etc/openvpn/server.conf
-	if [[ "$PROTOCOL" = 'UDP' ]]; then
-		echo "proto udp" >> /etc/openvpn/server.conf
-	elif [[ "$PROTOCOL" = 'TCP' ]]; then
-		echo "proto tcp" >> /etc/openvpn/server.conf
-	fi
+	echo "proto $(echo $PROTOCOL | tr '[:upper:]' '[:lower:]')" >> /etc/openvpn/server.conf
```
Comment (angristan) [inline openvpn-install.sh:518]: Nice 👏

**Comment by angristan** [comment]
Sorry I was in the exam period.
I left some questions.
Otherwise, LGTM. I will test the branch now

File: openvpn-install.sh (hunk 9 above)
Comment (sayem314) [reply openvpn-install.sh:264]: if [ $DNS -gt 1 ] && [ $DNS -lt 9 ]
    #  Returns true if both condition1 and condition2 hold true...

It’s shorter version of “&&” used inside.

Comment (sayem314) [reply openvpn-install.sh]: if [ $DNS -gt 1 ] && [ $DNS -lt 9 ]
    #  Returns true if both condition1 and condition2 hold true...

It’s shorter version of “&&” used inside.

File: openvpn-install.sh (hunk 10 above)
Comment (sayem314) [reply openvpn-install.sh:333]: It’s regex. ‘+’ means 1 or more and ‘$’ assert position at end of the string.

**Comment by sayem314** [comment]
> I will test the branch now

You should start using Travis CI

File: openvpn-install.sh (hunk 9 above)
Comment (angristan) [reply openvpn-install.sh:264]: Thanks

**Comment by angristan** [comment]
Thanks for you work 👍

File: openvpn-install.sh (hunk 10 above)
Comment (cezar97) [reply openvpn-install.sh]: So I'm not the only one that doesn't regex.

**Pull request merged**